### PR TITLE
Refactor TLS to use a TlsManager to obtain trust and key information,…

### PR DESCRIPTION
**Pull request by tomas-langer**
… retaining capability to reload TLS

### CODE REVIEW

File: common/tls/src/main/java/io/helidon/common/tls/TlsConfigDecorator.java
Comment (trentjeff) [inline common/tls/src/main/java/io/helidon/common/tls/TlsConfigDecorator.java:31]: Don't you think it would make more sense to leave it to the manager to decorate?  Seems like a more flexible design for what managers can/should do.

File: common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java
Comment (trentjeff) [inline common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java:57]: Why do we have to carry two of each? Seems conceptual to only carry a single one that may or may not be reloadable. Applies to the TrustManager also.

File: common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java
Comment (trentjeff) [inline common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java:96]: If SSLContext is coming out of the builder, do you want to remove it from the blueprint?  I had it so the manager->decorator was building the SSLContext that would get into the finalized config object.  I thought we wanted the SSLContext to be a long-lived instance, where only the underlying key and trust managers would mutate.

File: common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java
Comment (trentjeff) [inline common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java:118]: misspelled use

File: common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java
Comment (trentjeff) [inline common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java:289]: Also note that besides SSLContext, the secureRandom was capable of being set by the builder/decorator.  Seems like we are losing that here too if managers are not decorators.  With this new approach the secureRandom seems to be produced via `getInstance()` each time.

File: common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java
Comment (trentjeff) [inline common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java:274]: perhaps a better name would be to overload `initSslContext`

File: integrations/oci/tls-certificates/src/main/java/io/helidon/integrations/oci/tls/certificates/DefaultOciCertificatesTlsManager.java
Comment (trentjeff) [inline integrations/oci/tls-certificates/src/main/java/io/helidon/integrations/oci/tls/certificates/DefaultOciCertificatesTlsManager.java:87]: i would prefer to only realize these services if we are enabled as it was prior.

File: integrations/oci/tls-certificates/src/main/java/io/helidon/integrations/oci/tls/certificates/DefaultOciCertificatesTlsManager.java
Comment (trentjeff) [inline integrations/oci/tls-certificates/src/main/java/io/helidon/integrations/oci/tls/certificates/DefaultOciCertificatesTlsManager.java:187]: Seems we need a broader exception catch here since we are also downloading the pk from OCI which will throw a RuntimeException and not a KeyStoreException.

File: common/tls/src/main/java/io/helidon/common/tls/TlsConfigDecorator.java
Comment (tomas-langer) [reply common/tls/src/main/java/io/helidon/common/tls/TlsConfigDecorator.java:31]: No, as the SSLParameters are not the responsibility of the manager

File: common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java
Comment (tomas-langer) [reply common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java:57]: We cannot reload using a reloadable component (or we should not). The reason is to be able to provide the unwrapped X509 manager used to create this instance, and to have the reloadable instance, so we can reload when requested from outside (or from a descendant).

File: common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java
Comment (tomas-langer) [reply common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java:96]: From which blueprint? We can still set an explicit `SSLContext` on the builder of `TlsConfig` (and it is honored by the `ConfiguredTlsManager`). 
"SSLContext to be a long -lived instance": That is correct. `init(TlsConfig)` is called exactly once for the lifetime of a `Tls` instance, and the `SSLContext` instance is never changed after this method is invoked. I have also modified the OCI manager to honor this approach (it inits SSLContext just once on init, and calls `reload()` on sub-sequent changes to the certificates/keys

File: common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java
Comment (tomas-langer) [reply common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java:118]: please fix it once you merge this

File: common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java
Comment (tomas-langer) [reply common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java:289]: No, secure random is used from the configuration if configured by the user, and it uses the configured algorithm and provider (if user configures them).
The `RANDOM.get()` is a fallback if neither is configured (by the user).
Do you have a use case where the manager wants to use some other `SecureRandom`?

File: common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java
Comment (tomas-langer) [reply common/tls/src/main/java/io/helidon/common/tls/ConfiguredTlsManager.java:274]: Not really, as this is a method that is specific to the configured manager (and it does call the `initSslContext` as the last thing). I just did not want to pollute the `init(TlsConfig)` method with the code that is in this method.

File: common/tls/src/main/java/io/helidon/common/tls/TlsConfigDecorator.java
Comment (trentjeff) [reply common/tls/src/main/java/io/helidon/common/tls/TlsConfigDecorator.java:31]: I meant in general, not necessarily limiting to ssl params... anything on the builder.  This gives us a level of extensibility for also setting the secure random, and other configuration based upon the manager's impl.

File: integrations/oci/tls-certificates/src/main/java/io/helidon/integrations/oci/tls/certificates/DefaultOciCertificatesTlsManager.java
Comment (tomas-langer) [reply integrations/oci/tls-certificates/src/main/java/io/helidon/integrations/oci/tls/certificates/DefaultOciCertificatesTlsManager.java:87]: check the javadoc, this method is never called if the TLS is disabled.

File: integrations/oci/tls-certificates/src/main/java/io/helidon/integrations/oci/tls/certificates/DefaultOciCertificatesTlsManager.java
Comment (tomas-langer) [reply integrations/oci/tls-certificates/src/main/java/io/helidon/integrations/oci/tls/certificates/DefaultOciCertificatesTlsManager.java:187]: 1. you should not catch "any" runtime exception - some of the code throws our custom exception
2. you should never throw a `RuntimeException` type - we should either use a custom exception, or one of the predefined ones that make sense (this is part of our development guidelines)

If you want to wrap runtime exceptions from OCI (not sure this is really needed):
1. create a custom exception in our TLS module to mark our exceptions (when we fail to create secure random, keystore, factories etc.)
2. catch this exception and re-throw it without wrapping
3. catch a runtime exception and wrap it in your exception

**Review by trentjeff — APPROVED** [review]

**Pull request merged**
